### ISSUE TITLE
Deep hardening roadmap: adversarial evidence layer for paper-2 claims

### ISSUE THREAD

**Issue by omarespejel** [omarespejel/llm-provable-computer#153]
## Context

After Phase 37, the repo has a strong operational evidence chain:

- Phase 35: canonical recursive target manifest.
- Phase 36: source-bound verifier harness receipt over the Phase 35 target.
- Phase 37: heavier artifact-chain harness receipt from Phase 29 + Phase 30 through the derived Phase 36 receipt.
- Follow-up hardening: malformed Phase 37 commitment fields are rejected, and the malformed-commitment regression is in the local merge gate.

This is a serious base, but the next risk is not just a normal Rust bug. The next risk is **self-confirming verification**:

- the same code creates the artifacts,
- the same code verifies the artifacts,
- the paper describes the artifacts,
- and all three may share the same mistaken assumption.

The goal of this issue is to add a deeper assurance layer so the paper and implementation do not fool us into overclaiming.

## Goal

Create an adversarial evidence layer around the paper-facing Phase 29-37 surfaces.

The target standard is:

> Every important paper claim has independent evidence, negative controls, reproducible provenance, and an explicit non-claim boundary.

This should make the repo stronger for both research credibility and future industry-facing use.

### 1. Claim evidence matrix

Add a machine-readable claim map, for example:

- `docs/engineering/paper2-claim-evidence.yml`

For each paper-facing claim, record:

- claim text,
- implementation function(s),
- schema file(s),
- CLI command(s),
- artifact file(s),
- positive tests,
- negative/tamper tests,
- fuzz target or reason missing,
- Kani/formal harness or reason missing,
- evidence command,
- explicit non-claim boundary.

Then extend `scripts/paper/paper_preflight.py` so it fails if a paper-facing claim lacks evidence.

Example shape:

```yaml
- claim: Phase 37 recomputes the Phase 29 + Phase 30 through Phase 36 artifact chain.
  implementation:
    - src/stwo_backend/recursion.rs:phase37_prepare_recursive_artifact_chain_harness_receipt
    - src/stwo_backend/recursion.rs:verify_phase37_recursive_artifact_chain_harness_receipt_against_sources
  schema:
    - spec/stwo-phase37-recursive-artifact-chain-harness-receipt.schema.json
  positive_tests:
    - phase37_recursive_artifact_chain_harness_receipt_accepts_matching_sources
  negative_tests:
    - phase37_recursive_artifact_chain_harness_receipt_rejects_tampered_commitment
    - phase37_recursive_artifact_chain_harness_receipt_rejects_malformed_commitment_field
  non_claims:
    - Does not claim recursive proof verification.
    - Does not claim cryptographic compression.
```

### 2. Known-bad artifact corpus

Add a tracked corpus, for example:

- `tests/fixtures/known_bad/phase29_to_phase37/`

Include bad artifacts such as:

- malformed JSON,
- unknown fields,
- uppercase hex,
- short hex,
- non-hex strings,
- wrong version,
- wrong semantic scope,
- wrong backend,
- `recursive_verification_claimed=true`,
- `cryptographic_compression_claimed=true`,
- step count mismatch,
- boundary mismatch,
- public-input ordering mismatch,
- lookup commitment mismatch,
- self-consistent but semantically false artifacts where the outer commitment was recomputed.

The last category is especially important. A weak verifier rejects only a bad root commitment. A stronger verifier rejects a bad artifact even if the attacker recomputes the outer commitment.

### 3. Adversarial mutation generator

Add a generator, for example:

- `scripts/generate_bad_phase_artifacts.py`

It should take a valid Phase 29/30/31/32/33/34/35/36/37 chain and emit mutated variants with expected failures.

Example manifest entry:

```json
{
  "name": "phase37_wrong_phase35_commitment_recommitted",
  "input": "phase37_wrong_phase35_commitment_recommitted.json",
  "expected_error_contains": "phase35_recursive_target_manifest_commitment"
}
```

This gives systematic negative controls instead of only hand-written one-off tests.

### 4. Fuzz every public artifact loader

Add dedicated fuzz targets for public artifact loaders/parsers:

- Phase 29 contract JSON,
- Phase 30 envelope manifest JSON,
- Phase 31 bridge manifest JSON,
- Phase 32 statement contract JSON,
- Phase 33 public-input manifest JSON,
- Phase 34 shared-lookup manifest JSON,
- Phase 35 target manifest JSON,
- Phase 36 receipt JSON,
- Phase 37 receipt JSON,
- HF provenance manifest JSON,
- ONNX metadata manifest JSON.

Rule:

> Any public artifact parser must accept valid input, reject invalid input, and never panic.

### 5. Independent reference verifier

Add a small independent verifier, ideally under:

- `tools/reference_verifier/`

It should not reuse the main Rust structs. It should independently:

- parse JSON,
- check constants,
- canonicalize fields,
- recompute commitments,
- verify Phase 29 through Phase 37 commitments and source bindings.

This is the best way to catch common-mode errors between producer, verifier, and paper text.

### 6. Release evidence bundle and provenance verification

For paper or release checkpoints, produce a release evidence bundle with:

- git SHA,
- rustc/toolchain versions,
- schema hashes,
- artifact hashes,
- command logs,
- log SHA-256 digests,
- local merge-gate evidence,
- optional GitHub artifact attestation verification.

The repo already has HF provenance and external attestation binding. The next step is to actually verify release attestations and bind the verified digest into the release/provenance surface.

### 7. Mutation-testing survivor tracking

The repo already has a mutation-testing workflow. Make the output operational:

- run mutation testing on trusted-core verifier files before major paper/release milestones,
- record killed/survived/timeouts,
- add `docs/engineering/mutation-survivors.md`,
- either kill surviving mutants with new tests or document why they are equivalent/not useful.

### 8. Small Kani/formal harnesses for critical invariants

Do not try to formally verify the whole system. Add bounded formal checks for small invariants:

- length-prefixed commitment updates are unambiguous for bounded inputs,
- Phase 37 hash-format validation accepts only lowercase 64-char hex,
- valid receipts cannot have recursive/compression claim flags set,
- bounded public-input ordering preservation,
- parse/load wrappers return explicit errors rather than panicking on bounded malformed input.

All Kani results should be described as bounded checks, not full proofs.

### 9. Claim-language linter

Extend paper preflight with a simple overclaim linter.

Examples:

- reject `recursive proof` unless a nearby sentence also states the non-closure boundary,
- reject `proves model inference` unless the bounded statement is explicit,
- reject `semantic equivalence` unless `bounded` appears nearby,
- reject `supply-chain attestation` unless verified external attestation is implemented.

This prevents paper drift.

### 10. External red-team / reproducibility check

Before calling this industry-usable, ask someone else to implement a verifier from the paper/spec without reading the Rust verifier.

If they cannot reproduce Phase 37 verification from the docs, the docs/spec are not clear enough.

If their verifier disagrees with ours, we found a real issue.

## Suggested PR order

1. Claim evidence matrix + paper preflight enforcement.
2. Known-bad Phase 29-37 corpus + corpus runner.
3. Adversarial mutation generator for self-consistent bad artifacts.
4. Fuzz targets for Phase 29-37 public artifact loaders.
5. Independent reference verifier.
6. Release evidence bundle + verified attestation workflow.
7. Mutation survivor tracking.
8. Small Kani/formal invariant harnesses.
9. Claim-language linter.

## Acceptance criteria

- Paper-facing claims are mapped to code, tests, schemas, and explicit non-claims.
- Every Phase 29-37 public verifier has at least one valid path and multiple negative paths.
- Self-consistent bad artifacts are rejected, not just bad root commitments.
- Public artifact loaders have fuzz coverage or documented non-applicability.
- A reference verifier can independently validate the core Phase 29-37 commitments.
- Release evidence bundles bind commands, logs, schemas, artifacts, and git SHAs.
- The paper preflight fails on missing claim evidence or obvious overclaims.

## Non-goals

- Do not claim recursive proof closure until a real recursively verifiable compressed proof object exists.
- Do not claim full semantic equivalence across all runtimes/exporters.
- Do not claim complete supply-chain attestation unless external attestations are actually verified.
- Do not run expensive full CI/benchmarks on every PR by default; keep heavy validation intentional and evidence-backed.

## Sources and standards references

- NIST Secure Software Development Framework, SP 800-218: https://csrc.nist.gov/pubs/sp/800/218/final
- SLSA Build Track basics: https://slsa.dev/spec/v1.2/build-track-basics
- SLSA Build Provenance: https://slsa.dev/spec/v1.2/build-provenance
- SLSA Source Requirements: https://slsa.dev/spec/v1.2/source-requirements
- in-toto Attestation Framework: https://github.com/in-toto/attestation
- GitHub Artifact Attestations: https://docs.github.com/en/actions/concepts/security/artifact-attestations
- GitHub artifact attestation verification: https://docs.github.com/en/actions/how-tos/secure-your-work/use-artifact-attestations/use-artifact-attestations
- Rust Fuzz Book / cargo-fuzz: https://rust-fuzz.github.io/book/cargo-fuzz.html
- Miri: https://github.com/rust-lang/miri
- Kani Rust verifier docs: https://model-checking.github.io/kani/
- Kani loop unwinding / bounded proof notes: https://model-checking.github.io/kani/tutorial-loop-unwinding.html
- OpenSSF Scorecard: https://github.com/ossf/scorecard
- NIST Adversarial Machine Learning taxonomy, AI 100-2e2025: https://csrc.nist.gov/pubs/ai/100/2/e2025/final
- OWASP Top 10 for LLM Applications: https://owasp.org/www-project-top-10-for-large-language-model-applications/

## Benchmark, differential-reference, and approximation-budget addendum

The original roadmap focuses on adversarial artifacts, parser hardening, claim evidence, and provenance. Based on a wider survey of ZKML/zkLLM practice, it should also include a missing pillar: benchmark evidence, independent semantic reference checks, and explicit approximation budgets.

The extra risk is not only that a malformed artifact is accepted. It is also that the system can be internally consistent while proving, measuring, or publishing the wrong workload.

### 11. Differential reference execution

Add an independent reference-execution harness that does not share the VM/proof-artifact path.

Suggested files:

- `scripts/reference/run_reference_decode.py`
- `tests/fixtures/reference_cases/`
- `docs/engineering/differential-reference-plan.md`

For each paper-facing model/runtime semantic claim, compare the repo runtime or artifact output against at least one independent reference path, such as a small Python/NumPy implementation, a PyTorch/Hugging Face path, or an ONNX Runtime path where applicable.

This should record:

- model/input fixture hash,
- tokenizer or token fixture hash,
- runtime output,
- reference output,
- exact/approximate comparison rule,
- tolerated error budget,
- command used to reproduce the comparison.

This catches the common-mode failure where the producer, verifier, and paper all agree on a computation that is not actually the intended transformer computation.

### 12. Benchmark harness and methodology

Add reproducible benchmarks before paper-facing performance claims are made.

Suggested files:

- `benchmarks/README.md`
- `benchmarks/phase_chain_bench.py`
- `benchmarks/operator_bench.py`
- `benchmarks/reference_equivalence_bench.py`
- `benchmarks/results/`
- `docs/engineering/benchmark-methodology.md`

Metrics should include, when applicable:

- artifact generation time,
- verifier time,
- prover/preparation time,
- proof or receipt size,
- trace rows,
- symbolic constraints,
- peak RSS/VRAM,
- public input count,
- commitment count,
- lookup table size,
- hash count,
- schema validation time,
- known-bad corpus runner time,
- fuzz corpus size,
- mutation score.

Benchmark rules:

- no single-run timing claims,
- report at least p50 and p95 when repeated runs are feasible,
- include git SHA, hardware, OS, Rust/Python/toolchain versions, command line, and input artifact hashes,
- separate cold and warm runs when caches matter,
- never cite a benchmark number in the paper unless its JSON result and reproduction command are checked into an evidence bundle.

### 13. Approximation and quantization budget

If any approximate or quantized transformer operation is used, add an explicit budget instead of vague equivalence language.

Suggested files:

- `docs/engineering/approximation-budget.md`
- `tests/approximation/`
- `scripts/check_approximation_budget.py`

Track, as applicable:

- max/mean logit error,
- KL divergence,
- top-1/top-k agreement,
- perplexity delta,
- greedy exact-token match,
- token flips,
- accepted numerical tolerance by operator.

Paper preflight should reject claims like `semantic equivalence`, `preserves accuracy`, or `same model behavior` unless the claim points to bounded evidence and an explicit non-claim boundary.

### 14. Threat model and red-team matrix

Add a threat model that makes the hardening work auditable.

Suggested files:

- `docs/security/threat-model.md`
- `docs/security/red-team-matrix.md`

Include adversaries such as:

- malformed artifact producer,
- recomputed bad-artifact producer,
- model-substitution attacker,
- tokenizer-substitution attacker,
- quantization-substitution attacker,
- cached-response attacker,
- public-input-ordering attacker,
- backend-confusion attacker,
- schema-drift attacker,
- provenance-forgery attacker,
- paper-overclaim attacker.

Each adversary should map to at least one negative test, fuzz target, formal harness, benchmark/evidence check, or documented non-goal.

## Revised suggested PR order

1. Claim evidence matrix + paper preflight enforcement.
2. Threat model + red-team matrix.
3. Known-bad Phase 29-37 corpus + corpus runner.
4. Adversarial mutation generator for self-consistent bad artifacts.
5. Differential reference execution harness.
6. Benchmark harness + methodology.
7. Approximation/quantization budget.
8. Fuzz targets for Phase 29-37 public artifact loaders.
9. Independent reference verifier.
10. Release evidence bundle + verified attestation workflow.
11. Mutation survivor tracking.
12. Small Kani/formal invariant harnesses.
13. Claim-language linter.
14. External red-team/reproducibility check.

## Additional acceptance criteria

- Every paper-facing performance number has a reproducible benchmark JSON result.
- Every benchmark records git SHA, toolchain, hardware, command, and input artifact hashes.
- Every model/VM semantic claim has an independent reference execution case or a documented non-goal.
- Every approximate/quantized operation has an explicit error budget.
- Every end-to-end claim has positive and negative controls.
- No single-run timing number may be cited as evidence.
- No `semantic equivalence`, `preserves accuracy`, or similar claim passes paper preflight without bounded evidence.

## Additional research references

- NANOZK: Accelerating Folding-based ZKML Inference with Nested Accumulation: https://arxiv.org/abs/2603.18046
- zkLLM: Zero Knowledge Proofs for Large Language Models: https://hongyanz.github.io/publications/CCS_zkLLM.pdf
- zkLLM implementation repository: https://github.com/jvhs0706/zkllm-ccs2024
- zkLLM benchmark repository: https://github.com/jvhs0706/zkllm-benchmark
- zkGPT: Zero-Knowledge Proofs for Secure and Verifiable AI: https://eprint.iacr.org/2025/1184
- Jolt Atlas: https://arxiv.org/abs/2602.17452
- ZKML: An Optimizing System for ML Inference in Zero-Knowledge Proofs: https://ddkang.github.io/papers/2024/zkml-eurosys.pdf
- MLCommons benchmarks: https://mlcommons.org/benchmarks/
- MLPerf Inference documentation: https://docs.mlcommons.org/inference/
- EleutherAI Language Model Evaluation Harness: https://github.com/EleutherAI/lm-evaluation-harness

**Comment by omarespejel** [comment]
Issue #153 hardening execution update, 2026-04-17:

Merged PRs:

- #157 threat model and research-backed hardening plan: https://github.com/omarespejel/provable-transformer-vm/pull/157
- #158 differential reference / approximation-budget harness: https://github.com/omarespejel/provable-transformer-vm/pull/158
- #159 benchmark evidence harness and methodology: https://github.com/omarespejel/provable-transformer-vm/pull/159
- #160 paper overclaim / claim-language preflight linter: https://github.com/omarespejel/provable-transformer-vm/pull/160

Merge-gate evidence produced locally:

- `target/local-hardening/pr-157-c032246e514cf89fe98869410bdf3a3413aee51a/evidence.json`
- `target/local-hardening/pr-158-b460df0f73e91e6768a409d8b12dc5ac3c2767b9/evidence.json`
- `target/local-hardening/pr-159-9676963af00e05cd50c19281d22dc8360af9bce9/evidence.json`
- `target/local-hardening/pr-160-b784288c1424e4eb45bc29824893e2767177d600/evidence.json`

Review discipline applied:

- Addressed useful Qodo / CodeRabbit comments before merge.
- Resolved only stale threads after verifying the latest code or docs covered them.
- Used the local merge gate with `--quiet-seconds 360 --wait --merge`; #158 waited through the enforced quiet window after the final Qodo/CodeRabbit activity before merging.

Implemented safeguards now in `main`:

- Threat-model matrix for fooling-ourselves risks and validation scope.
- Deterministic differential-reference fixture/check harness.
- Fail-closed approximation-budget checker and negative fixtures.
- Non-default `top_k` reference fixture coverage.
- Benchmark manifest runner with dry-run/run evidence, input hashing rules, timeout behavior, path validation, and parser tests.
- Paper preflight overclaim linter with regression coverage for missing/empty paper directories and scoped claim contexts.
